src/Streamer.cxx

```diff
@@ -84,225 +80,251 @@ int FileWriter::Streamer::get_topic_partitions(const std::string &topic) {
+void FileWriter::Streamer::initialize_streamer(
+    FileWriter::Streamer::Options &filewriter_options) {
+  filewriter_options.erase(std::remove_if(filewriter_options.begin(),
+                                          filewriter_options.end(),
+                                          [&](option_t &item) -> bool {
+                                            return this->set_streamer_opt(item);
+                                          }),
+                           filewriter_options.end());
+  if (!filewriter_options.empty()) {
+    for (auto &item : filewriter_options) {
+      LOG(3, "Unknown option: {} [{}]", item.first, item.second);
+    }
   }
+}
 
-  if (!topic_name.empty()) {
-    status_.emplace("status.metadata", 1);
-    if (get_metadata() != 0) {
-      LOG(1, "Unable to retrieve metadata");
-      status_["status.metadata"] = -1;
-    }
-    status_.emplace("status.topic_partition", 1);
-    if (get_topic_partitions(topic_name) < 0) {
-      LOG(1, "Unable to build TopicPartitions structure");
-      status_["status.topic_partition"] = -1;
-    }
-    for (auto &i : _tp) {
-      i->set_offset(_offset.value());
-    }
+FileWriter::Streamer::Streamer(const std::string &broker,
+                               const std::string &topic_name,
+                               FileWriter::Streamer::Options kafka_options,
+                               FileWriter::Streamer::Options filewriter_options)
+    : run_status_{SEC::not_initialized} {
 
-    RdKafka::ErrorCode err = _consumer->assign(_tp);
-    if (err) {
-      LOG(0, "Failed to subscribe to {} ", topic_name);
-      status_["status.consumer"] = -1;
-    }
-  } else {
-    LOG(0, "Topic required");
-    status_["status.consumer"] = -1;
+  if (topic_name.empty() || broker.empty()) {
+    LOG(0, "Broker and topic required");
+    run_status_ = SEC::not_initialized;
+    return;
   }
-  if (get_offset_boundaries().value()) {
-    LOG(3, "Unable to determine lower and higher offset in topic {}",
-        topic_name);
-    status_["status.metadata"] = -1;
+  kafka_options.push_back(
+      FileWriter::Streamer::option_t{"metadata.broker.list", broker});
+  kafka_options.push_back(
+      FileWriter::Streamer::option_t{"api.version.request", "true"});
+  kafka_options.push_back(
+      FileWriter::Streamer::option_t{"group.id", topic_name});
+
+  connect_ = std::thread([&] {
+    this->connect(topic_name, kafka_options, filewriter_options);
+    return;
+  });
+
+  std::unique_lock<std::mutex> lk(connection_lock_);
+  connection_ready_.wait(lk, [&] { return this->ready_.load(); });
+}
+
+FileWriter::Streamer::~Streamer() { close_stream(); }
+
+void FileWriter::Streamer::connect(
+    const std::string topic_name, FileWriter::Streamer::Options kafka_options,
+    FileWriter::Streamer::Options filewriter_options) {
+  std::lock_guard<std::mutex> lock(guard_);
+
+  std::lock_guard<std::mutex> lk(connection_lock_);
+  connection_ready_.notify_all();
+  ready_ = true;
```

Comment:
I'm not sure what the `ready_` flag is for, just wanted to ask if it should be maybe rather set before `notify_all()`?

Reply:
:white_check_mark:
I also changed their names in order to be at least a little descriptive